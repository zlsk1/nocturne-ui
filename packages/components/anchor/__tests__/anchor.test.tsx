import { nextTick, ref } from 'vue'
import { mount } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import Anchor from '../src/anchor.vue'
import AnchorItem from '../src/anchor-item.vue'

let id = 0

const getHash = () => `#anchor-item-${id++}`

describe('Anchor.vue', () => {
  test('snapshot', async () => {
    const hash = getHash()
    const wrapper = mount({
      props: ['direction'],
      render() {
        return (
          <Anchor>
            <AnchorItem href={hash}>{hash}</AnchorItem>
          </Anchor>
        )
      }
    })

    wrapper.setProps({
      direction: 'vertical'
    })
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      direction: 'horizontal'
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      direction: 'vertical'
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      direction: 'horizontal'
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('scroll after click', async () => {
    const hash = getHash()
    const wrapper = mount(
      () => (
        <>
          <div id={hash.replace('#', '')}>hash</div>
          <Anchor>
            <AnchorItem href={hash}>{hash}</AnchorItem>
          </Anchor>
        </>
      ),
      {
        attachTo: 'body'
      }
    )
    const scrollSpy = vi
      .spyOn(window, 'scrollTo')
      .mockImplementation(() => undefined)
    await nextTick()
    wrapper.find(`a[href="${hash}"]`).trigger('click')
    await nextTick()
    expect(scrollSpy).toHaveBeenCalledTimes(1)
  })

  test('scrollTo method', async () => {
    const hash = getHash()
    const anchorRef = ref<InstanceType<typeof Anchor> | null>(null)
    mount(
      () => (
        <>
          <div id={hash.replace('#', '')}>hash</div>
          <Anchor ref={anchorRef}>
            <AnchorItem href={hash}>{hash}</AnchorItem>
          </Anchor>
        </>
      ),
      {
        attachTo: 'body'
      }
    )
    const scrollSpy = vi
      .spyOn(window, 'scrollTo')
      .mockImplementation(() => undefined)
    await nextTick()
    anchorRef.value?.handleScrollTo(hash)
    await nextTick()
    expect(scrollSpy).toHaveBeenCalledTimes(1)
  })

  test('change event', async () => {
    const hash1 = getHash()
    const hash2 = getHash()
    const wrapper = mount(() => (
      <Anchor>
        <AnchorItem href={hash1}>{hash1}</AnchorItem>
        <AnchorItem href={hash2}>{hash2}</AnchorItem>
      </Anchor>
    ))
    wrapper.find(`a[href="${hash1}"]`).trigger('click')
    wrapper.find(`a[href="${hash2}"]`).trigger('click')
    expect(wrapper.findComponent(Anchor).emitted('change')).toEqual([
      [hash1],
      [hash2]
    ])
  })

  test('click event', async () => {
    const hash = getHash()
    const wrapper = mount(() => (
      <Anchor>
        <AnchorItem href={hash}>{hash}</AnchorItem>
      </Anchor>
    ))
    wrapper.find(`a[href="${hash}"]`).trigger('click')

    expect(
      (wrapper.findComponent(Anchor).emitted('click') as any)[0][0]
    ).toBeInstanceOf(MouseEvent)
    expect((wrapper.findComponent(Anchor).emitted('click') as any)[0][1]).toBe(
      hash
    )
  })
})
