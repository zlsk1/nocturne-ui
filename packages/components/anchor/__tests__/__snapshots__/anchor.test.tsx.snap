// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Anchor.vue > snapshot 1`] = `
"<div class="n-anchor"><span class="n-anchor__marker"></span>
  <div class="n-anchor__content">
    <div class="n-anchor__item"><a href="#anchor-item-0" class="">
        <div class="n-anchor__item__text">#anchor-item-0</div>
      </a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`Anchor.vue > snapshot 2`] = `
"<div class="n-anchor"><span class="n-anchor__marker"></span>
  <div class="n-anchor__content">
    <div class="n-anchor__item"><a href="#anchor-item-0" class="">
        <div class="n-anchor__item__text">#anchor-item-0</div>
      </a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`Anchor.vue > snapshot 3`] = `
"<div class="n-anchor"><span class="n-anchor__marker"></span>
  <div class="n-anchor__content">
    <div class="n-anchor__item"><a href="#anchor-item-0" class="">
        <div class="n-anchor__item__text">#anchor-item-0</div>
      </a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`Anchor.vue > snapshot 4`] = `
"<div class="n-anchor"><span class="n-anchor__marker"></span>
  <div class="n-anchor__content">
    <div class="n-anchor__item"><a href="#anchor-item-0" class="">
        <div class="n-anchor__item__text">#anchor-item-0</div>
      </a>
      <!--v-if-->
    </div>
  </div>
</div>"
`;
